### PR TITLE
Defer creation of HTTP listener sockets to @AcceptRequests

### PR DESCRIPTION
**Pull request by johngmyers**

### CODE REVIEW

**Review by gwittel — APPROVED** [review]
LGTM

File: http-server/src/main/java/com/proofpoint/http/server/HttpServerInfo.java [hunk 1]
```diff
@@ -165,4 +178,116 @@ private static ServerSocketChannel createChannel(InetAddress address, int port,
...
+
+    static class LazyChannel implements ChannelHolder
+    {
+        private final String scheme;
+        private final String hostname;
+        private final String externalHostname;
+        private final InetAddress address;
+        private final int port;
+        private final int acceptQueueSize;
+        private boolean isAdmin;
```
Comment (therontock) [inline http-server/src/main/java/com/proofpoint/http/server/HttpServerInfo.java:220]: shouldn't admin be final as well?

File: http-server/src/main/java/com/proofpoint/http/server/HttpServerInfo.java [hunk 2]
```diff
@@ -165,4 +178,116 @@ private static ServerSocketChannel createChannel(InetAddress address, int port,
...
+            synchronized (this) {
+                if (channel == null) {
+                    try {
+                        ServerSocketChannel newChannel = ServerSocketChannel.open();
+                        newChannel.socket().setReuseAddress(true);
+                        newChannel.socket().bind(new InetSocketAddress(address, port), acceptQueueSize);
+                        channel = newChannel;
+                    }
+                    catch (IOException e) {
+                        throw new UncheckedIOException(e);
```
Comment (therontock) [inline http-server/src/main/java/com/proofpoint/http/server/HttpServerInfo.java:252]: Should call newChannel.close() if newChannel != null.

Comment (johngmyers) [reply http-server/src/main/java/com/proofpoint/http/server/HttpServerInfo.java:252]: `newChannel` is out of scope here. While we could do a nested try block, that's getting overly complicated for what is in effect going to shut down the process.

**Pull request merged**